lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.cpp

```diff
@@ -0,0 +1,64 @@
+#include "lib/Analysis/MulDepthAnalysis/MulDepthAnalysis.h"  // from @llvm-project
+
+#include "lib/Dialect/Openfhe/IR/OpenfheOps.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"  // from @llvm-project
+#include "llvm/include/llvm/Support/Debug.h"   // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"    // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"        // from @llvm-project
+#include "mlir/include/mlir/IR/Visitors.h"     // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"    // from @llvm-project
+
+#define DEBUG_TYPE "openfhe-muldepth-analysis"
+
+namespace mlir {
+namespace heir {
+namespace mul_depth_analysis {
+
+void MulDepthAnalysis::visitOperation(
+    Operation *op, ArrayRef<const MulDepthLattice *> operands,
+    ArrayRef<MulDepthLattice *> results) {
+  llvm::TypeSwitch<Operation &>(*op)
+      .Case<openfhe::MulOp, openfhe::MulPlainOp, openfhe::MulConstOp,
```

Comment:
MulPlain and MulConst I think should not consume multiplicative depth. Or rather, if we want to track those it depends on the const/plain value on the other side, and that requires a more fine-grained noise analysis. I'd say leave them out. cc @AlexanderViand-Intel

Reply:
I might be wrong about this part because I believe I don't fully understand the characteristics of noise.

I agree that having MulPlain and MulConst consume multiplicative depth is a very conservative approach: https://openfhe.discourse.group/t/multiplication-depth-in-bfv/487/7

However, if we leave these out in this analysis, couldn't it cause a problem in programs that contain too many MulPlain/MulConst ops(though I'm not sure if there are any practical use cases for this) due to a lower multiplicative depth than actual?

How about keeping the current conservative approach for now, and then changing it when we can calculate multiplicative depth considering MulPlain/MulConst through a more fine-grained noise analysis?

Reply:
Mh, this is tricky. If we want to use the MulDepth Analysis for Relin insertion, we'd have to ignore them, while for "rescale" they do count. 
And of course they do add some smaller but noticeable noise so things that heavily use them (e.g., ML inference) would probably get wrong parameters.

I guess if the goal is to get a conservative parameter estimation, we should probably count them.

Reply:
I remember when the @Maokami first posted the issue I went through a rabbit hole trying to see what other libraries or compilers do. Apparently OpenFHE also counts plaintext-ciphertext multiplications the same as ciphertext multiplications (https://openfhe.discourse.group/t/multiplication-depth-in-bfv/487/5), and decided that we should be conservative for now too.